### PR TITLE
fix(sql): symbol cannot be set to CACHE immediately after setting it to NOCACHE

### PR DESCRIPTION
**Pull request by nwoolmer**
In `SymbolMapWriter`, there is a condition to determine whether the symbol is or isn't cached:

https://github.com/questdb/questdb/blob/753485fb4427b9317a98921c59031d300d4cc2da/core/src/main/java/io/questdb/cairo/SymbolMapWriter.java#L200-L202

This checks whether or not there are values in-memory in a cache.

In `TableWriter`, there is a `changeCacheFlag` method:

https://github.com/questdb/questdb/blob/753485fb4427b9317a98921c59031d300d4cc2da/core/src/main/java/io/questdb/cairo/TableWriter.java#L817-L829

This is used to toggle the flag on or off.

Starting with `CACHE` enabled, the `SymbolMapWriter` is loaded and cache initialised.

When setting it to `NOCACHE`, `SymbolMapWriter` will run `updateCacheFlag`:

https://github.com/questdb/questdb/blob/753485fb4427b9317a98921c59031d300d4cc2da/core/src/main/java/io/questdb/cairo/SymbolMapWriter.java#L269-L271

This writes the new flag value. However, it does not null `cache` or dispose of `SymbolMapWriter`.

When trying to reinitialise this to `CACHE`, the condition is checked. But since the `cache` object has not been nulled, it is a noop, and the `CACHE` flag is not written to disk.

A first attempt to resolve was to change `isCached()` to rely on the disk value:

image

This will then break a number of tests in `FullFwdPartitionFrameCursorTest`:

image

**The fix**

The fix is to add a separate cache flag, and to bump truncate version in txWriter to make sure all readers and writers see the change.

### CODE REVIEW

**Comment by questdb-butler** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Please update the PR title to match this format:
`type(subType): description`

Where `type` is one of:
`feat`, `fix`, `chore`, `docs`, `style`, `refactor`, `perf`, `test`, `ci`, `revert`

And: `subType` is one of:
`build`, `sql`, `log`, `mig`, `core`, `ilp`, `pgwire`, `http`, `conf`, `ui`, `wal`

For Example:

```
perf(sql): improve pattern matching performance for SELECT sub-queries
```
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 0f074a5103947156ad1a62bb17a5a794c11e90b4
</p>

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|

File: core/src/main/java/io/questdb/cairo/SymbolMapWriter.java [hunk 1]
```diff
@@ -53,6 +53,7 @@ public class SymbolMapWriter implements Closeable, MapWriter {
     private final int maxHash;
     private final int symbolCapacity;
     private final SymbolValueCountCollector valueCountCollector;
+    private boolean cached_flag;
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cairo/SymbolMapWriter.java:56]: nit: field names in Java code should be in camelCase, not snake_case

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**